### PR TITLE
Don't set background to black if ~/.background-image not present

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change
In https://github.com/NixOS/nixpkgs/pull/49492 @mkaito mentions their usecase of running `nitrogen --restore` to set the background image, which could be implemented with e.g. `services.xserver.windowManager.i3.extraSessionCommands`, but only if the code for `~/.background-image` wouldn't run after that, which sets a black background, which was done because of https://github.com/NixOS/nixpkgs/pull/25101.

This PR gets rid of setting the background to black again, which I think was the wrong solution to that problem.

This can be tested with a `configuration.nix` like
```nix
{ config, pkgs, ... }: {
  services.xserver = {
    enable = true;
    displayManager.auto = {
      enable = true;
      user = "demo";
    };
    windowManager.i3 = {
      enable = true;
      extraSessionCommands = ''
        ${pkgs.feh}/bin/feh --bg-max ${config.boot.loader.grub.splashImage}
      '';
    };
  };

  users.users.demo = {
    initialPassword = "";
    isNormalUser = true;
  };
}
```

which when ran with
```bash
$ nix-build nixos --arg configuration ./configuration.nix -A vm
$ result/bin/run-nixos-vm
```

results in the background image being set:

which wasn't possible like this previously

Ping @mkaito @worldofpeace @jtojnar @hedning 

Note: The first commit is just a cleanup

### CODE REVIEW

File: nixos/modules/services/x11/desktop-managers/default.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/default.nix]: I don't think this should have a fallback code path, in other desktop environments I noticed it being a solid black wallpaper when it should be the texture used from the window manager. Feels extra.

Comment (Infinisil) [reply nixos/modules/services/x11/desktop-managers/default.nix]: Yeah probably just removing that is a better solution. I'm mainly wondering then why it's there in the first place.

Comment (worldofpeace) [reply nixos/modules/services/x11/desktop-managers/default.nix]: > Yeah probably just removing that is a better solution. I'm mainly wondering then why it's there in the first place.

```
  # If desktop manager `d' isn't capable of setting a background and
  # the xserver is enabled, `feh' or `xsetroot' are used as a fallback.
```

So I guess by reasoning above ^ makes this a proper feature.

It was introduced in https://github.com/NixOS/nixpkgs/commit/852813689a335399e30f83e0659666f3e312716e which is an invalid motivation IMHO. We shouldn't compensate for a desktop-manager that can't set a wallpaper in 2020??
And it conflicts with the majority that can.

Comment (Infinisil) [reply nixos/modules/services/x11/desktop-managers/default.nix]: `git blame` gave me the answer: https://github.com/NixOS/nixpkgs/pull/25101

Comment (worldofpeace) [reply nixos/modules/services/x11/desktop-managers/default.nix]: I'm actually wondering why I ran into that in the first place.
It happened with desktopManagers that didn't use `services.xserver.desktopManager.session` to have the `bgSupport` attribute.

Comment (Infinisil) [reply nixos/modules/services/x11/desktop-managers/default.nix]: So yeah, I think that PR was the wrong solution to that problem. And I can't reproduce the issue from that PR. Using the current version of the PR without the fallback branch and the following config:
```nix
{
  services.xserver = {
    enable = true;
    displayManager.sddm.enable = true;
    windowManager.i3.enable = true;
  };

  users.users.demo = {
    initialPassword = "";
    isNormalUser = true;
  };
}
```

I get the following after logging in:

Comment (jtojnar) [reply nixos/modules/services/x11/desktop-managers/default.nix]: Hmm, maybe i3 started resetting the root window but I am pretty sure I encountered this issue in some DM/WM combination when testing https://github.com/NixOS/nixpkgs/pull/53843

Comment (jtojnar) [reply nixos/modules/services/x11/desktop-managers/default.nix]: It was LightDM & XMonad: https://github.com/NixOS/nixpkgs/pull/53843#issuecomment-565480115

**Comment by Infinisil** [comment]
@primeos Could you test this PR to see if the issue you had in https://github.com/NixOS/nixpkgs/pull/25101 still occurs with it?

**Comment by primeos** [comment]
@Infinisil unfortunately not, I switched to Sway and don't have my previous SDDM + i3 configuration anymore (I could create a new test setup of course, but since you've already tested it with SDDM + i3 it shouldn't add any value / give different results).

I also assume that i3 started resetting the root window in the meantime, like @jtojnar mentioned, which sounds like a good idea in general.

And I agree that defaulting to a black background was just another hack / incomplete solution. However, I'm also not sure if `extraSessionCommands` is really the right place to configure the background (IIRC these commands are executed before the WM). In general I also don't like the extra logic for `~/.background-image` in general (IIRC undocumented, not very flexible, hardcoded path, etc), but that's another discussion.

Anyway, I don't have any objections and this PR LGTM (it might still cause a few regressions that result in something similar to https://github.com/NixOS/nixpkgs/pull/25101, but a black background is also not always optimal).

**Comment by Infinisil** [comment]
Cool, thanks. I think the most important thing here is that the default behavior is to do nothing. So if the user doesn't have a `.background-image`, NixOS doesn't do anything special.

**Review by mkaito — APPROVED** [review]
Looks good!

**Comment by jluttine** [comment]
What is the correct way to set the background to black now? (Using i3 with SDDM.)

**Comment by jluttine** [comment]
I suppose running this explicitly?

```
${pkgs.xorg.xsetroot}/bin/xsetroot -solid black
```

**Comment by primeos** [comment]
@jluttine do you still see the behaviour described in https://github.com/NixOS/nixpkgs/pull/25101 (SDDM login screen as background) or is there a different issue now?

Regarding the fix: Yes, running `xsetroot -solid black` should be the best solution IMO (`man i3` seems to suggest setting this in `~/.xsession` - therefore setting it in `extraSessionCommands` should be fine as well (and as `exec` in the i3 config file it should of course also be fine)).

**Comment by jluttine** [comment]
@primeos yeah, that looks like the issue I had after this PR was merged. But indeed using `xsetroot` explicitly fixes that. :+1:

**Comment by primeos** [comment]
Ok, thanks for the update :) Then this behaviour might be hardware/driver dependant :o

Not sure what we should do in this case / how many users are affected.
i3 can probably do nothing about this (as that should break `~/.xsession` scripts and our `extraSessionCommands` without extra hacks) and the only thing we can do is probably to add `${pkgs.xorg.xsetroot}/bin/xsetroot -solid black` to `nixos/modules/services/x11/window-managers/i3.nix` right before `${cfg.extraSessionCommands}` (but this would only cover i3 and kinda restore the previous behaviour with the exception that it now runs before `extraSessionCommands` but it might still run after `~/.xsession` (haven't looked into this)).

**Pull request merged**
